File: Test/AdventureWorksLegacyModel/Human Resources/Department.cs
// Copyright Naked Objects Group Ltd, 45 Station Road, Henley on Thames, UK, RG9 1AT
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.

using System;
using System.ComponentModel.DataAnnotations;
using Legacy.Types;
using NakedObjects;


namespace AdventureWorksModel {

    [Bounded]
    [Immutable]
    [LegacyType]
    public class Department  {
        #region Life Cycle Methods
        public virtual void Persisting()  => ModifiedDate.DateTime = DateTime.Now;

        public virtual void Updating()  => Persisting();

        #endregion

        [NakedObjectsIgnore]
        public virtual short DepartmentID { get; set; }

        #region Name
        internal string mappedName;
        internal TextString myName;

        [MemberOrder(1)]
        public virtual TextString Name => myName ??= new TextString(mappedName, s => mappedName = s);
        #endregion

        #region GroupName
        internal string mappedGroupName;
        private TextString cachedGroupName;


        [MemberOrder(2)]
        public virtual TextString GroupName => cachedGroupName ??= new TextString(mappedGroupName, s => mappedGroupName = s );
        #endregion

        #region ModifiedDate
        internal DateTime mappedModifiedDate;
        private TimeStamp myModifiedDate;

        [MemberOrder(99)]
        [Disabled]
        [ConcurrencyCheck]
        public virtual TimeStamp ModifiedDate => myModifiedDate ??= new TimeStamp(mappedModifiedDate, s => mappedModifiedDate = s);
        #endregion

        public Title title() => Name.Title();

    }
}

File: Test/AdventureWorksLegacyModelVB/Human Resources/Department.vb
Namespace AW.Types

    '<Bounded>
    Partial Public Class Department

        Public Property DepartmentID As Integer  'Not visible on UI

#Region "Name"
        Public mappedName As String
        Friend myName As TextString

        '<MemberOrder(1)>
        Public ReadOnly Property Name As TextString
            Get
                Return If(myName, New TextString(mappedName, Function(v) mappedName = v))
            End Get
        End Property

        Public Sub AboutName(a As FieldAbout, Name As TextString)
            Select Case a.TypeCode
                Case AboutTypeCodes.Name
                Case AboutTypeCodes.Usable
                Case AboutTypeCodes.Valid
                Case AboutTypeCodes.Visible
            End Select
        End Sub
#End Region

#Region "GroupName"
        Public mappedGroupName As String
        Friend myGroupName As TextString

        '<MemberOrder(2)>
        Public ReadOnly Property GroupName As TextString
            Get
                Return If(myGroupName, New TextString(mappedGroupName, Function(v) mappedGroupName = v))
            End Get
        End Property
#End Region

#Region "ModifiedDate"
        Public mappedModifiedDate As Date
        Friend myModifiedDate As TimeStamp

        '<MemberOrder(99)>
        Public ReadOnly Property ModifiedDate As TimeStamp
            Get
                Return If(myModifiedDate, New TimeStamp(mappedModifiedDate, Function(v) mappedModifiedDate = v))
            End Get
        End Property

        Public Sub AboutModifiedDate(a As FieldAbout, ModifiedDate As TimeStamp)
            Select Case a.TypeCode
                Case AboutTypeCodes.Usable
                    a.Usable = False
            End Select
        End Sub
#End Region

        Public Function Title() As Title
            Return New Title(Name)
        End Function
    End Class
End Namespace